File: HOT100/11.二分查找/154.寻找旋转排序数组中的最小值-ii.cpp
/*
 * @lc app=leetcode.cn id=154 lang=cpp
 *
 * [154] 寻找旋转排序数组中的最小值 II
 */
#include "tools.h"
using namespace std;
// @lc code=start
class Solution {
public:
    int findMin(vector<int>& nums) {
        int l = 0, r = nums.size() - 1;
        while (l < r)
        {
            int mid = l + r >> 1;
            if (nums[mid] > nums[r]) l = mid + 1;
            else if (nums[mid] < nums[r]) r = mid;
            else r --;
        }
        return nums[r];
    }
};
// @lc code=end

File: HOT100/HOT100p1/11.二分查找/154.寻找旋转排序数组中的最小值-ii.py
#
# @lc app=leetcode.cn id=154 lang=python3
#
# [154] 寻找旋转排序数组中的最小值 II
#
from mytools import *
# @lc code=start
class Solution:
    def findMin(self, nums: List[int]) -> int:
        l, r = 0, len(nums) - 1
        while l < r:
            mid = l + r >> 1
            if nums[mid] > nums[r]: l = mid + 1
            elif nums[mid] < nums[r]: r = mid
            else: r -= 1
        return nums[r]
# @lc code=end

